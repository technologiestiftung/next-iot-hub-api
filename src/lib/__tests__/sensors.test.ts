/* eslint-disable jest/require-top-level-describe */
/* eslint-disable jest/no-hooks */
// Copyright (c) 2021 Technologiestiftung Berlin & Fabian MorÃ³n Zirfas
//
// This software is released under the MIT License.
// https://opensource.org/licenses/MIT

import { definitions } from "@technologiestiftung/stadtpuls-supabase-definitions";
import {
  buildServerOpts,
  closePool,
  connectPool,
  createSensor,
  createSensors,
  maxRows,
  sensorsEndpoint,
  signupUser,
  truncateTables,
} from "../../__test-utils";
import buildServer from "../server";
const sensorSnapshotDescription = {
  data: [
    {
      altitude: expect.any(Number),
      category: {
        id: expect.any(Number),
        description: expect.any(String),
        name: expect.any(String),
      },
      author: {
        name: expect.any(String),
        display_name: null,
      },
      connection_type: "http",
      created_at: expect.any(String),
      description: null,
      external_id: null,
      icon_id: null,
      id: 1,
      latitude: expect.any(Number),
      location: null,
      longitude: expect.any(Number),
      name: "test",
    },
  ],
};
beforeAll(async () => {
  await connectPool();
});
beforeEach(async () => {
  await truncateTables();
});
// eslint-disable-next-line jest/no-hooks
afterAll(async () => {
  await truncateTables();
  await closePool();
});
describe(`all ${sensorsEndpoint} tests`, () => {
  test("should be rejected due to wrong query string", async () => {
    const server = buildServer(buildServerOpts);
    const response = await server.inject({
      method: "GET",
      url: `${sensorsEndpoint}?wrong=param`,
    });
    expect(response.statusCode).toBe(400);
    expect(response.json()).toMatchInlineSnapshot(`
      Object {
        "error": "Bad Request",
        "message": "querystring should NOT have additional properties",
        "statusCode": 400,
      }
    `);
  });
  test("list of all sensors GET (empty)", async () => {
    const server = buildServer(buildServerOpts);
    const response = await server.inject({
      method: "GET",
      url: sensorsEndpoint,
    });
    expect(response.statusCode).toBe(200);
    expect(response.json()).toStrictEqual({
      data: [],
    });
  });
  test("list of all sensors HEAD", async () => {
    const server = buildServer(buildServerOpts);
    const response = await server.inject({
      method: "HEAD",
      url: sensorsEndpoint,
    });

    expect(response.statusCode).toBe(200);
    expect(response.body).toBe("");
    expect(response.headers["content-range"]).toBeDefined();
    expect(response.headers["range-unit"]).toBeDefined();
  });

  test("get list of all sensors GET > 0", async () => {
    const server = buildServer(buildServerOpts);
    const user = await signupUser();
    await createSensor({ user_id: user.id, name: "test" });
    const response = await server.inject({
      method: "GET",
      url: sensorsEndpoint,
    });
    expect(response.statusCode).toBe(200);
    expect(response.json().data).toHaveLength(1);
    expect(response.json().data[0]).toMatchSnapshot(
      sensorSnapshotDescription.data[0]
    );
  });

  test(`get list of all sensors GET > ${maxRows} should have pagination`, async () => {
    const server = buildServer(buildServerOpts);
    const user = await signupUser();
    await createSensors(user.id, maxRows * 2);
    const response = await server.inject({
      method: "GET",
      url: sensorsEndpoint,
    });
    const json = response.json<{
      url: string;
      nextPage: string;
      data: Omit<definitions["sensors"], "user_id">[];
    }>();
    const lastItem = json.data[json.data.length - 1];
    expect(json).toMatchSnapshot({
      data: expect.any(Array),
      nextPage: `/api/v3/sensors?offset=${maxRows}&limit=${maxRows}`,
    });
    expect(lastItem.id).toBe(maxRows);
    expect(response.statusCode).toBe(200);

    // make another request
    const response2 = await server.inject({
      method: "GET",
      url: json.nextPage,
    });
    const json2 = response2.json<{
      nextPage: string;
      data: Omit<definitions["sensors"], "user_id">[];
    }>();

    expect(response2.statusCode).toBe(200);
    expect(json2.data[0].id).toBe(maxRows + 1);
  });
});

describe(`single ${sensorsEndpoint}/:sensorId tests`, () => {
  test("should give us 404 response", async () => {
    const server = buildServer(buildServerOpts);
    const response = await server.inject({
      method: "HEAD",
      url: `${sensorsEndpoint}/1`,
    });
    expect(response.statusCode).toBe(404);
    expect(response.body).toMatchInlineSnapshot(
      `"{\\"statusCode\\":404,\\"error\\":\\"Not Found\\",\\"message\\":\\"Not Found\\"}"`
    );
  });
  test("should give us HEAD response", async () => {
    const server = buildServer(buildServerOpts);
    const user = await signupUser();
    const sensor = await createSensor({ user_id: user.id, name: "test" });
    const response = await server.inject({
      method: "HEAD",
      url: `${sensorsEndpoint}/${sensor.id}`,
    });
    expect(response.statusCode).toBe(200);
    expect(response.body).toBe("");
    expect(response.headers["content-range"]).toBeDefined();
    expect(response.headers["range-unit"]).toBeDefined();
  });
  test("should give us the sensor response", async () => {
    const server = buildServer(buildServerOpts);
    const user = await signupUser();
    const sensor = await createSensor({ user_id: user.id, name: "test" });
    const response = await server.inject({
      method: "GET",
      url: `${sensorsEndpoint}/${sensor.id}`,
    });
    expect(response.statusCode).toBe(200);
    expect(response.json()).toMatchSnapshot(sensorSnapshotDescription);
  });
});
