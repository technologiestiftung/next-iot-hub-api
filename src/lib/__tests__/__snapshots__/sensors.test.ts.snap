// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`all /api/v3/sensors tests get list of all sensors GET > 0 1`] = `
Object {
  "altitude": Any<Number>,
  "author": Object {
    "display_name": null,
    "name": Any<String>,
  },
  "category": Object {
    "description": Any<String>,
    "id": Any<Number>,
    "name": Any<String>,
  },
  "connection_type": "http",
  "created_at": Any<String>,
  "description": null,
  "external_id": null,
  "icon_id": null,
  "id": 1,
  "latitude": Any<Number>,
  "location": null,
  "longitude": Any<Number>,
  "name": "test",
}
`;

exports[`all /api/v3/sensors tests get list of all sensors GET > 1000 should have pagination 1`] = `
Object {
  "data": Any<Array>,
  "nextPage": "/api/v3/sensors?offset=1000&limit=1000",
}
`;

exports[`single /api/v3/sensors/:sensorId tests should give us the sensor response 1`] = `
Object {
  "data": Array [
    Object {
      "altitude": Any<Number>,
      "author": Object {
        "display_name": null,
        "name": Any<String>,
      },
      "category": Object {
        "description": Any<String>,
        "id": Any<Number>,
        "name": Any<String>,
      },
      "connection_type": "http",
      "created_at": Any<String>,
      "description": null,
      "external_id": null,
      "icon_id": null,
      "id": 1,
      "latitude": Any<Number>,
      "location": null,
      "longitude": Any<Number>,
      "name": "test",
    },
  ],
}
`;
